python/tank/descriptor/io_descriptor/downloadable.py

```diff
@@ -0,0 +1,139 @@
+        Downloads the data represented by the descriptor into the primary bundle
+        cache path.
+
+        It does so in a two step process. First, by downloading it to
+        a temporary bundle cache path (typically in a 'tmp/<uuid>' directory
+        in the bundle cache path), then, by moving the data to the primary bundle
+        cache path for that descriptor. This helps to guard against multiple
+        processes attempting to download the same descriptor simultaneously.
+        """
+
+        # Return if the descriptor exists locally.
+        if self.exists_local():
+            return
+
+        # cache into a temporary location
+        temporary_path = self._get_temporary_cache_path()
+
+        # move into primary location
+        target = self._get_primary_cache_path()
+
+        # ensure that the parent directory of the target is present.
+        # make sure we guard against multiple processes attempting to create it simultaneously.
+        try:
+            filesystem.ensure_folder_exists(os.path.dirname(target))
+        except Exception as e:
+            if not os.path.exists(os.path.dirname(target)):
+                log.error("Failed to create parent directory %s: %s" % (os.path.dirname(target), e))
+                raise TankDescriptorError("Failed to create parent directory %s: %s" % (os.path.dirname(target), e))
+
+        try:
+            # attempt to download the descriptor to the temporary path.
+            log.debug("Downloading %s to the temporary download path %s." % (self, temporary_path))
+            self._download_local(temporary_path)
+        except Exception as e:
+            # something went wrong during the download, remove the temporary files.
+            log.error("Failed to download into path %s: %s. Attempting to remove it."
+                      % (temporary_path, e))
+            filesystem.safe_delete_folder(temporary_path)
+            raise TankDescriptorError("Failed to download into path %s: %s" % (temporary_path, e))
+
+        success = False
+        log.debug("Attempting to move descriptor %s from temporary path %s to target path %s." % (
+            self, temporary_path, target)
+        )
+
+        try:
+            # atomically rename the directory temporary_path to the target.
+            os.rename(temporary_path, target)
+            success = True
+            log.debug("Successfully moved the downloaded descriptor to target path: %s." % target)
+        except Exception as e:
+            # if the target path does not already exist, it something else might have gone wrong.
+            if not os.path.exists(target):
+                log.error("Failed to move descriptor from the temporary path %s to " +
+                          "the bundle cache %s: %s" % (temporary_path, target, e))
+                raise TankError("Failed to move descriptor from the temporary path %s to " +
+                                "the bundle cache %s: %s" % (temporary_path, target, e))
```

Comment:
Maybe you could use `try`, `except`, `else`,  `finally` here so you don't need a `success` flag? The `else` part of the `try` will be executed only if the `try` didn't fail.